internal/idutils/external_id.go

```diff
@@ -18,26 +18,34 @@ package idutils
 
 import (
 	"encoding/base64"
-	"fmt"
+	"github.com/dynatrace/dynatrace-configuration-as-code/pkg/config/v2/coordinate"
+	"strings"
 )
 
-// GenerateExternalID generates the externalID for settings 2.0 objects based on the schema, and ID.
-// The result of the function is pure.
-// Max length for the external ID is 500
-func GenerateExternalID(schema, id string) string {
+// GenerateExternalID returns a string that can be used as an external ID for a Settings 2.0 object
+func GenerateExternalID(coordinate coordinate.Coordinate) string {
+	if coordinate.Type == "" || coordinate.ConfigId == "" {
+		panic("Settings schema id and config id needs to be set to generate a valid external id for settings 2.0 objects")
+	}
+	if coordinate.Project == "" {
+		return generateExternalID(coordinate.Type, coordinate.ConfigId)
+	}
+	return generateExternalID(coordinate.Project, coordinate.Type, coordinate.ConfigId)
+}
+
+func generateExternalID(parts ...string) string {
 	const prefix = "monaco:"
-	const format = "%s$%s"
 	const externalIDMaxLength = 500
 
-	formattedID := fmt.Sprintf(format, schema, id)
-	encodedID := base64.StdEncoding.EncodeToString([]byte(formattedID))
-
-	encodedIDMaxLength := externalIDMaxLength - len(prefix)
-	if len(encodedID) > encodedIDMaxLength {
-		encodedID = encodedID[encodedIDMaxLength:]
+	rawID := strings.Join(parts, "$")
+	if strings.HasSuffix(rawID, "$") {
```

Comment:
Wouldn't this only apply if the configID actually ends in $ ? And then we'd produce possibly wrong/overlapping external ID.

Join places the separator between elements AFAIK - so for no or one element it won't append a $

Reply:
true that